patroni/ctl.py

```diff
@@ -1372,3 +1373,222 @@ def format_pg_version(version):
+        cur.execute("SELECT name, current_setting(name) FROM pg_settings\
+                     WHERE context <> 'internal' \
+                     AND source IN ('configuration file', 'command line', 'environment variable') \
+                     AND category <> 'Write-Ahead Log / Recovery Target' \
+                     AND setting <> '(disabled)' \
+                     OR name IN ('hba_file', 'ident_file', 'config_file', \
+                                 'data_directory', \
+                                 'listen_addresses', 'port', \
+                                 'max_connections', 'max_worker_processes', 'max_wal_senders', \
+                                 'max_replication_slots', 'max_locks_per_transaction', 'max_prepared_transactions', \
+                                 'hot_standby', \
+                                 'wal_level', 'wal_log_hints', \
+                                 'wal_keep_segments', 'wal_keep_size', \
+                                 'track_commit_timestamp');")
+
+        # adjust values
+        for p, v in cur.fetchall():
+            if p == 'data_directory':
+                config['postgresql']['data_dir'] = v
+            elif p in ('archive_command', 'restore_command', 'archive_cleanup_command',
+                       'recovery_end_command', 'ssl_passphrase_command'):
+                # write commands to the local config due to security implications
+                if not config['postgresql']['parameters']:
+                    config['postgresql']['parameters'] = {}
+                config['postgresql']['parameters'][p] = v
+            else:
+                config['bootstrap']['dcs']['postgresql']['parameters'][p] = v
+
+    conn.close()
+
+    if not config['postgresql']['bin_dir']:
+        # obtain bin_dir of the running instance
+        postmaster_pid = None
+        try:
+            with open(f"{config['postgresql']['data_dir']}/postmaster.pid", 'r') as f:
+                postmaster_pid = f.readline()
+                if not postmaster_pid:
+                    raise PatroniCtlException('Failed to obtain postmaster pid from postmaster.pid file')
+                postmaster_pid = int(postmaster_pid.strip())
+        except OSError as e:
+            raise PatroniCtlException(f'Error while reading postmaster.pid file: {e}')
+        try:
+            import psutil
+            config['postgresql']['bin_dir'] = psutil.Process(postmaster_pid).exe()
+        except psutil.NoSuchProcess:
+            raise PatroniCtlException('Obtained postmaster pid doesn\'t exist')
+
+    config['postgresql']['connect_address'] = f'{parsed_dsn["host"]}:{parsed_dsn["port"]}'
+    listen_addresses = config['bootstrap']['dcs']['postgresql']['parameters']['listen_addresses']
+    port = config['bootstrap']['dcs']['postgresql']['parameters']['port']
+    config['postgresql']['listen'] = f'{listen_addresses}:{port}'
+
+    try:
+        with open(f"{config['postgresql']['data_dir']}/pg_hba.conf", 'r') as f:
```

Comment:
pg_hba.conf could be outside of $PGDATA.
Also, we need to have something similar for pg_ident.conf

Reply:
meh, with pg_read_file it was easer:)